### PR TITLE
Enforce limit on maximum number of label application rules that can be created per tenant

### PR DESCRIPTION
**Pull request by ankitchoudhary111**

### CODE REVIEW

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 1]
```diff
@@ -47,6 +57,25 @@ public void createLabelApplicationRule(
...
+        for (LabelApplicationRule rule : labelApplicationRules) {
+          if (rule.getData().getLabelAction().hasDynamicLabelExpression()) {
+            dynamicLabelApplicationRules = dynamicLabelApplicationRules + 1;
+          }
+        }
+        if (dynamicLabelApplicationRules >= maxLabelApplicationRuleAllowed) {
+          LOG.error(
+              "Failed to create Label Application Rule because the limit ({}) for the number of dynamic label application rules allowed has already been reached!",
+              maxLabelApplicationRuleAllowed);
+          responseObserver.onCompleted();
```
[the remark this answers is not in the document]
Comment (saxenakshitiz) [reply label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:75]: Status.RESOURCE_EXHAUSTED can be returned in this case

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 2]
```diff
@@ -47,6 +57,25 @@ public void createLabelApplicationRule(
...
+            this.labelApplicationRuleStore.getAllObjects(requestContext).stream()
+                .map(ConfigObject::getData)
+                .collect(Collectors.toUnmodifiableList());
+        int dynamicLabelApplicationRules = 0;
+        for (LabelApplicationRule rule : labelApplicationRules) {
+          if (rule.getData().getLabelAction().hasDynamicLabelExpression()) {
+            dynamicLabelApplicationRules = dynamicLabelApplicationRules + 1;
+          }
+        }
+        if (dynamicLabelApplicationRules >= maxLabelApplicationRuleAllowed) {
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:71]: Should be `>` check only

**Review by saxenakshitiz — COMMENTED** [review]
1. Add a unit test for this
2. Add limit in application.conf

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java (hunk 2 above)
Comment (ankitchoudhary111) [reply label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:71]: If the create request wants to create a rule with dynamic label then the count of already existing dynamic label application rules must be strictly less than the max_allowed limit. right?... hence the `>=` condition.

Comment (saxenakshitiz) [reply label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:72]: Makes sense. Missed the current rule part.

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 3]
```diff
@@ -23,14 +24,24 @@
...
+import org.slf4j.LoggerFactory;
 
 public class LabelApplicationRuleConfigServiceImpl
     extends LabelApplicationRuleConfigServiceGrpc.LabelApplicationRuleConfigServiceImplBase {
   private final IdentifiedObjectStore<LabelApplicationRule> labelApplicationRuleStore;
   private final LabelApplicationRuleValidator requestValidator;
+  private final int maxLabelApplicationRuleAllowed;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(LabelApplicationRuleConfigServiceImpl.class);
+  private final int MAX_LABEL_APPLICATION_RULE_CONSTANT = 100;
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:37]: static final. Also move all static final fields to top of fields declaration

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 4]
```diff
@@ -47,6 +58,22 @@ public void createLabelApplicationRule(
     try {
       RequestContext requestContext = RequestContext.CURRENT.get();
       this.requestValidator.validateOrThrow(requestContext, request);
+      if (request.getData().getLabelAction().hasDynamicLabelExpression()) {
+        List<LabelApplicationRule> labelApplicationRules =
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:62]: below code can be changed to -
```
this.labelApplicationRuleStore.getAllObjects(requestContext)
              .stream()
              .map(configObject -> configObject.getData().getData().getLabelAction())
              .filter(LabelApplicationRuleData.Action::hasDynamicLabelKey)
              .count();
```

**Comment by saxenakshitiz** [comment]
Resolve all comments that are addressed. Also add a unit test.

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 5]
```diff
@@ -26,11 +27,17 @@
...
   private final IdentifiedObjectStore<LabelApplicationRule> labelApplicationRuleStore;
   private final LabelApplicationRuleValidator requestValidator;
+  private final int maxLabelApplicationRuleAllowed;
 
   public LabelApplicationRuleConfigServiceImpl(
-      Channel configChannel, ConfigChangeEventGenerator configChangeEventGenerator) {
+      Channel configChannel, Config config, ConfigChangeEventGenerator configChangeEventGenerator) {
+    this.maxLabelApplicationRuleAllowed =
+        config.hasPath("max.label.application.rules.per.tenant")
+            ? config.getInt("max.label.application.rules.per.tenant")
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:39]: nit: Define constant for `max.label.application.rules.per.tenant`

File: helm/templates/config-service-config.yaml [hunk 6]
```diff
@@ -22,6 +22,7 @@ data:
         }
       }
       publish.change.events = {{ .Values.configServiceConfig.publishChangeEvents }}
+      max.label.application.rules.per.tenant = {{ .Values.configServiceConfig.maxLabelApplicationRulesPerTenant }}
```
Comment (saxenakshitiz) [inline helm/templates/config-service-config.yaml:25]: You will need to add maxLabelApplicationRulesPerTenant in values.yaml as well

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 7]
```diff
@@ -47,6 +54,22 @@ public void createLabelApplicationRule(
...
       RequestContext requestContext = RequestContext.CURRENT.get();
       this.requestValidator.validateOrThrow(requestContext, request);
+      if (request.getData().getLabelAction().hasDynamicLabelKey()
+          || request.getData().getLabelAction().hasDynamicLabelExpression()) {
+        int dynamicLabelApplicationRules =
+            (int)
+                this.labelApplicationRuleStore.getAllObjects(requestContext).stream()
+                    .map(configObject -> configObject.getData().getData().getLabelAction())
+                    .filter(
+                        ruleDataAction ->
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:64]: nit: `action` is sufficient

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java [hunk 8]
```diff
@@ -47,6 +54,22 @@ public void createLabelApplicationRule(
     try {
       RequestContext requestContext = RequestContext.CURRENT.get();
       this.requestValidator.validateOrThrow(requestContext, request);
+      if (request.getData().getLabelAction().hasDynamicLabelKey()
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java:57]: Please extract this part to a separate method

File: helm/templates/config-service-config.yaml (hunk 6 above)
Comment (ankitchoudhary111) [reply helm/templates/config-service-config.yaml:25]: https://github.com/Traceableai/helm-values/pull/1005

Comment (saxenakshitiz) [reply helm/templates/config-service-config.yaml:25]: I meant in values.yaml in this repo `helm/values.yaml`

**Review by saxenakshitiz — APPROVED** [review]

**Pull request merged**